### PR TITLE
add Model.from_files, change bulk -> predict

### PR DESCRIPTION
**Pull request by joelgrus**
(1) add a `Model.from_files` static method that instantiates a trained model from the experiment config (and has some optional overrides).
(2) add a `Params.from_file()` static method to abstract out some logic I was using in a lot of different places
(3) made both server and tests explicit about which `Predictable`s they want to load and serve, rather than having a "default" collection
(4) renamed `bulk` -> `predict` and fixed it to work properly. now it takes an experiment config and some input data. (I set up "default" predictors per model, it's possible we might want to allow an override in `predict` but I think it's good for now)

### CODE REVIEW

File: allennlp/commands/predict.py [hunk 1]
```diff
@@ -1,29 +1,28 @@
...
+from allennlp.common.params import Params
+from allennlp.service.predictors import Predictor
 
 def add_subparser(parser: argparse._SubParsersAction) -> argparse.ArgumentParser:  # pylint: disable=protected-access
     description = '''Run the specified model against a JSON-lines input file.'''
     subparser = parser.add_parser(
-            'bulk', description=description, help='Run a model in bulk.')
-    subparser.add_argument('model', type=str, help='the name of the model to run')
+            'predict', description=description, help='Use a trained model to make predictions.')
+    subparser.add_argument('config_file', type=str, help='the configuration file that trained the model')
```
Comment (schmmd) [inline allennlp/commands/predict.py]: Should we refer to them as "experiment configuration files"?

File: allennlp/common/params.py [hunk 2]
```diff
@@ -184,6 +185,15 @@ def __check_is_dict(self, new_history, value):
             return Params(value, new_history)
         return value
 
+    # TODO(joelgrus): should this be pyhocon instead of json?
```
Comment (matt-gardner) [inline allennlp/common/params.py]: Yes, I think so.

File: allennlp/commands/predict.py (hunk 1 above)
Comment (matt-gardner) [inline allennlp/commands/predict.py]: Or "training config".

**Comment by joelgrus** [comment]
ok, I think this is ready for review

File: allennlp/models/model.py [hunk 3]
```diff
@@ -87,3 +91,55 @@ def get_metrics(self, reset: bool = False) -> Dict[str, float]:
...
+
+        weights_file = weights_file or os.path.join(serialization_prefix, 'best.th')
+
+        # Load vocabulary from file
+        vocab_dir = os.path.join(serialization_prefix, 'vocabulary')
+        vocab = Vocabulary.from_files(vocab_dir)
+
+        model = Model.from_params(vocab, config.get('model'))
+        model_state = torch.load(weights_file, map_location=device_mapping(cuda_device))
+        model.load_state_dict(model_state)
```
Comment (matt-gardner) [inline allennlp/models/model.py]: Can we just call `model.cuda(cuda_device)` or `model.cpu()` here, and remove the need for `map_location` entirely?

Comment (joelgrus) [inline allennlp/models/model.py]: `torch.load(weights_file)` gives an error if you load gpu-trained weights on cpu (or vice versa) without using `map_location`

```
AssertionError: Torch not compiled with CUDA enabled
```

Comment (matt-gardner) [inline allennlp/models/model.py]: Huh...  Do we need to call `model.cuda()` / `model.cpu()` _after_ this, then?  I'm just thinking that this is likely the solution to one of the bugs we talked about earlier.

Comment (joelgrus) [inline allennlp/models/model.py]: I can definitely add that

Comment (matt-gardner) [inline allennlp/models/model.py]: What I don't know is if that would then mask an underlying problem, or if it's actually the right thing to do.  I guess a "save and load" test that actually does things this way should tell us?

**Comment by matt-gardner** [comment]
Hmm, one other question: how do you feel about something like `load_saved_model` instead of `from_files`?  I know the docstring makes it clear that this is loading something that's been trained, but it'd be nice if the name of the function made that obvious, too.  It gives a bigger distinction between `from_params` and `from_files`.

**Comment by matt-gardner** [comment]
Even just `load()` would probably be enough, because the arguments make it obvious what it's loading from.

**Comment by joelgrus** [comment]
ok, I renamed `from_files()` to `load()`. also I added a test in `SrlPredictorTest` that loads both a gpu and cpu trained models and compares the results. A more ideal test would train a GPU model, save it, and load it as a CPU model, but we still don't have a way to run GPU tests. Also, conceptually that test doesn't *belong* in SrlPredictor, but it was easiest to write using the `Predictor` machinery.

**Comment by matt-gardner** [comment]
Thanks for doing that.  Don't worry about it in this PR, but I think the right way to test this is to modify [this code](https://github.com/allenai/allennlp/blob/6c0ceb9bfeaf7bd18f7bc9d83120cdc48705db2c/allennlp/common/testing/test_case.py#L85) to use the `Model.load()` functionality that you've just implemented.

**Pull request merged**